### PR TITLE
Language Integrated Query In C#

### LINKED ISSUE #2995: Language-Integrated Query (LINQ) in C#
### Introduction paragraph (2-3 paragraphs):
Language-Integrated Query (LINQ) is the name for a bunch of advancements dependent on the reconciliation of inquiry capacities straightforwardly into the C# language. Customarily, inquiries against information are communicated as basic strings without type checking at aggregate time or IntelliSense support.

In this article , I will deeply look on how LINQ works eg how one can generate sequences using LINQ , how to write methods that can be easily used in LINQ queries amongst others.

### Key takeaways: 
1. What is LINQ?
2.  Generating  sequences with LINQ.
3.  Writing  methods that can be easily used in LINQ queries.
4. Distinguishing  between eager and lazy evaluation.
5. Locating areas in your code where the LINQ queries might run into performance issues .

### References: 
N/A

### PR DESCRIPTION
**Pull request by amosnjoro**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`
This closes #2995
> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by hectorkambow** [comment]
@paulodhiambo can we get a status update on this PR?? 

cc @amosnjoro - not sure if you are working on the plagiarism report.

**Comment by amosnjoro** [comment]
@hectorkambow I am working on it

**Comment by ahmadmardeni1** [comment]
plagiarism check done ✅

File: content/articles/language-integrated-query-in-c#/index.md.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    return sequence;
+}
+```
+In the above code, do not forget to add the input-output extension so that the program will execute.
+```c#
+using System.IO;
+```
+### Conclusion
+In conclusion, we have got to see some working functionalities of the LINQ in C# and how they are implemented in say a program or solving some real-life problems.
```
Comment (paulodhiambo) [inline content/articles/language-integrated-query-in-c#/index.md.md:156]: Change to "In conclusion, we have got to see some working functionalities of the LINQ in C# and how to implement them in a program or solve real-life problems."

File: content/articles/language-integrated-query-in-c#/index.md.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+```c#
+shuffle = shuffle.Skip(26).InterleaveSequenceWith(shuffle.Take(26));
+```
+When you execute the program once more, you will note that the deck 'takes 52 iterations to reshuffle itself. This is contributed to by several factors. You may address one of the key causes of this reduction in performance which is the inefficient lazy evaluation. Note that we used a LINQ query to create the original deck. The three queries of LINQ on the previous deck are used to generate each shuffle. All of this occurs at a glacial speed. This indicates that the sequence is repeated each time it is requested. By the moment you reach the fifty second iteration, you will have regenerated the initial deck severaly i.e more than once. Let's make a log to show how this works. Then you'll take care of it.
+
+Type or copy the method below into your Extensions.cs file. This extension method results into the creation of a new file which is named as debug.log in your project directory and logs the query that is presently being run. This extension method can be used to indicate the completion of a query.
```
Comment (paulodhiambo) [inline content/articles/language-integrated-query-in-c#/index.md.md:137]: change to "Type or copy the method below into your `Extensions.cs` file. This extension method creates a new file named.log in your project directory and logs the query that is presently being running. We can use this extension method to indicate the completion of a query."

**Comment by hectorkambow** [comment]
@amosnjoro 
Good afternoon and thank you for submitting your topic to the EngEd program and jumping on your revisions. 
As a rule of thumb - we can only accept articles (pull requests) that are fully polished and complete articles - articles that are ready to be published.
 
All revisions and edits should be completed on your own forked repo (as to not take up room in the queue). 

@paulodhiambo  if it seems this PR is still quite a bit aways from being completed. We will be closing this PR to allow @amosnjoro the time to revise and polish the article to be published. 

If closed - once it is completed (and ready to be published) go ahead and please re-submit as a new PR (and you can add the duplicate label) and ping the reviewer and myself. 
 
👍

**Comment by amosnjoro** [comment]
@hectorkambow I am working on the fixes but the main challenge was that my device had a serious problem but I have fixed it

File: content/articles/language-integrated-query-in-c#/index.md.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
+
+### Introduction
+The term "Language-Integrated Query" (LINQ) refers to a group of technologies that incorporate query capabilities directly into the C# programming language. Querying data has traditionally been plain text, with no type checking or IntelliSense assistance at build time. It's used to access data from objects, data sets, SQL Server, and XML, among other things.
+
+### Generating sequences with LINQ
+Make sure the below three lines are at the very top the Program.cs file produced by dotnet new console before you begin. For our program to run, the lines below should be in the program so that the results shall be obtained as expected
```
Comment (paulodhiambo) [inline content/articles/language-integrated-query-in-c#/index.md.md:6]: Before this point, please explain how to create the application.

**Review by paulodhiambo — CHANGES_REQUESTED** [review]

File: content/articles/language-integrated-query-in-c#/index.md.md (hunk 3 above)
Comment (paulodhiambo) [reply content/articles/language-integrated-query-in-c#/index.md.md]: @amosnjoro you're saying we add the three lines to program.cs file, how did we get to that program.cs file? Explain how to create the application and reach a point of having program.cs file.

Comment (amosnjoro) [reply content/articles/language-integrated-query-in-c#/index.md.md]: The first line of code which is using System means that you are using the System library in your project or program and that it  gives you some important functions and classes like WriteLine function and the Console class.

The second line of code which is using System.Collections.Generics  Contains interfaces and classes that define generic collections, which allow users to create strongly typed collections that provide better type safety and performance than non-generic strongly typed collections

The third line of code which is  using System.Linq provides classes and interfaces that support queries that use  (LINQ).

In short they act as preproccessor directives when dealing with LINQ. For example when connecting a window form to a database one have to include the line using System.Data.Sqlclient for it to work.

Comment (paulodhiambo) [reply content/articles/language-integrated-query-in-c#/index.md.md]: @amosnjoro When someone reads your article, you start by saying we add the 3 lines into program.cs file. Assume the reader is a beginner, he/she will be wondering where that program.cs file is located.

Comment (amosnjoro) [reply content/articles/language-integrated-query-in-c#/index.md.md]: What I meant by doing so is that when one is writing a program in c# you have to include the using System line of code and when in the same program you go ahead and introduce a linq concept and that will prompt you to add the other lines of code so that the program may be perfect

**Comment by hectorkambow** [comment]
@paulodhiambo is this PR in tech review? I saw that we had requested some changes from the author?

File: content/articles/language-integrated-query-in-c#/index.md.md (hunk 3 above)
Comment (paulodhiambo) [reply content/articles/language-integrated-query-in-c#/index.md.md]: @amosnjoro please check this article [Azure functions in C#](https://www.section.io/engineering-education/using-azure-functions-and-csharp-to-create-serverless-apis/). It explains how to create the application before adding any block of code.

**Comment by ahmadmardeni1** [comment]
Any update? @amosnjoro

**Comment by amosnjoro** [comment]
I think in order to avoid any confusion, I can just fix the lines of code in the program itself so that I can avoid the confusion that can be raised.
I can do that @ahmadmardeni1. Please allow me to do so.

**Comment by amosnjoro** [comment]
@ahmadmardeni1 I will highly appreciate your feedback

**Comment by ahmadmardeni1** [comment]
Hey @amosnjoro Sorry for the delay.
Sure thing, go ahead and fix the code. But what feedback do you want from me?
cc @paulodhiambo

**Comment by amosnjoro** [comment]
Thanks @ahmadmardeni1 for your reply. I just wanted to hear from you if I would fix the lines of code into the program

**Comment by amosnjoro** [comment]
@ahmadmardeni1 I have done the changes. You can have a look now.

**Comment by paulodhiambo** [comment]
> @ahmadmardeni1 I have done the changes. You can have a look now.

Good, I'm doing a final review.

**Comment by amosnjoro** [comment]
@paulodhiambo any problem that have risen please?

**Comment by amosnjoro** [comment]
@paulodhiambo  I have fixed the grammar, you can have a look now

**Comment by paulodhiambo** [comment]
@amosnjoro articles are always written in third person view, not the second person view. I am wondering why you reverted all the changes I made to your article.

**Comment by amosnjoro** [comment]
@paulodhiambo I am much sorry I thought I was supposed to fix it.  I can undo them

**Comment by amosnjoro** [comment]
@paulodhiambo  I have done as it was, sorry for the mistake.

**Comment by LinusMuema** [comment]
Hi @amosnjoro 
Please check your hero image to fit the required dimensions
Screenshot from 2021-09-28 21-24-51
.
You had checked that the photo had the required size but on rendering it shows up as shown

**Comment by amosnjoro** [comment]
Fine @LinusMuema

**Comment by amosnjoro** [comment]
@LinusMuema  I have fixed the hero image to the correct dimensions.

**Comment by LinusMuema** [comment]
Hi @amosnjoro 
Thank you for the article. Please go through the review comments and changes to improve your future articles. Also, you had uploaded your image to the wrong directory. Please check on that too.

cc: @paulodhiambo
